### ISSUE TITLE
Root btrfs volume visible under overlays in all docker containers

### ISSUE THREAD

**Issue by ComradeP** [docker/docker#10216]
**Description of problem**: Running Debian-packaged docker 1.3.3, on a btrfs filesystem (and hence defaulting to the btrfs backend).

Any container I create (including during `docker build`) has my host's root btrfs volume mounted as the basis of its root; the overlay filesystems are on top of it, but in particular any file from my root volume that's not explicitly deleted in an overlay is visible in the container.

This is particularly annoying for several common aspects of `/etc`, for example `/etc/apt/sources.list.d`, which adds my own sources into the `apt-get` configuration of any image I build, and `/etc/ld.so.conf.d`, which messes with the dynamic linker configuration in the running container.

`docker version`:
```
Client version: 1.3.3
Client API version: 1.15
Go version (client): go1.3.3
Git commit (client): d344625
OS/Arch (client): linux/amd64
Server version: 1.3.3
Server API version: 1.15
Go version (server): go1.3.3
Git commit (server): d344625
```

`docker info`:
```
Containers: 51
Images: 417
Storage Driver: btrfs
Execution Driver: native-0.2
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
WARNING: No memory limit support
WARNING: No swap limit support
```

`uname -a`:
```
Linux $HOSTNAME 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt4-1 (2015-01-16) x86_64 GNU/Linux
```

**Environment details (AWS, VirtualBox, physical, etc.):** Physical Debian Sid, with a btrfs volume for root and one for `/home`.

**How reproducible:** Always.

**Steps to Reproduce:**
1. `docker pull debian` (currently fetches image ID `4d6ce913b130`).
2. `docker run --rm -t -i debian bash -l`
3. In the container: `cat /etc/apt/sources.list.d/spotify.list` (which exists, since it exists on my root file system).

**Actual Results:** The contents of `/etc/apt/sources.list.d/spotify.list` from the host.

**Expected Results:** A file-not-found error, as the image doesn't contain this file.

**Additional info:**
```
### within the running container
# mount 
/dev/sdc5 on / type btrfs (rw,relatime,compress=lzo,ssd,space_cache)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev type tmpfs (rw,nosuid,mode=755)
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=666)
sysfs on /sys type sysfs (ro,nosuid,nodev,noexec,relatime)
/dev/sdc5 on /etc/resolv.conf type btrfs (rw,relatime,compress=lzo,ssd,space_cache)
/dev/sdc5 on /etc/hostname type btrfs (rw,relatime,compress=lzo,ssd,space_cache)
/dev/sdc5 on /etc/hosts type btrfs (rw,relatime,compress=lzo,ssd,space_cache)
devpts on /dev/console type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
proc on /proc/sys type proc (ro,nosuid,nodev,noexec,relatime)
proc on /proc/sysrq-trigger type proc (ro,nosuid,nodev,noexec,relatime)
proc on /proc/irq type proc (ro,nosuid,nodev,noexec,relatime)
proc on /proc/bus type proc (ro,nosuid,nodev,noexec,relatime)
tmpfs on /proc/kcore type tmpfs (rw,nosuid,mode=755)
# lsblk 
NAME                          MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                             8:0    0 465.8G  0 disk 
|-sda1                          8:1    0   1.5G  0 part 
|-sda2                          8:2    0 251.4G  0 part 
`-sda3                          8:3    0  17.6G  0 part 
sdb                             8:16   0 931.5G  0 disk 
|-sdb1                          8:17   0  29.8G  0 part [SWAP]
|-sdb2                          8:18   0     1K  0 part 
`-sdb5                          8:21   0 901.7G  0 part 
  |-vg0-root (dm-0)           254:0    0   7.5G  0 lvm  
  |-vg0-var (dm-1)            254:1    0   7.5G  0 lvm  
  |-vg0-usr (dm-2)            254:2    0  14.5G  0 lvm  
  |-vg0-usr--local (dm-3)     254:3    0   7.5G  0 lvm  
  |-vg0-tmp (dm-4)            254:4    0   7.5G  0 lvm  
  |-vg0-srv (dm-5)            254:5    0   7.5G  0 lvm  
  |-vg0-home (dm-6)           254:6    0 476.9G  0 lvm  
  `-vg0-btrfs (dm-7)          254:7    0   256G  0 lvm  
sdc                             8:32   0 931.5G  0 disk 
|-sdc1                          8:33   0  29.8G  0 part [SWAP]
|-sdc2                          8:34   0     1K  0 part 
`-sdc5                          8:37   0 901.7G  0 part /

**Comment by crosbymichael** [comment]
Thanks for the detailed report.  I'll look into this soon.

**Comment by gdm85** [comment]
May I suggest a ``security`` label for this?

**Comment by ewindisch** [comment]
I've failed to reproduce, but it could be a problem with my environment.

You have a single btrfs filesystem for the entire OS?

**Comment by ComradeP** [comment]
Yes, `/dev/sdc5` is the btrfs physical volume, and it has subvolumes which are mounted at `/`, `/tmp` and `/home`. Here's the relevant part of `/etc/fstab`:

```
/dev/sdc5 /               btrfs   defaults,compress=lzo,subvol=@           0       1
/dev/sdc5 /home           btrfs   defaults,compress=lzo,subvol=@home       0       2
/dev/sdc5 /tmp            btrfs   defaults,compress=lzo,subvol=@tmp        0       2
```

What can I do to help reproduce? This is still biting me, 100% reliably.

**Comment by gdm85** [comment]
Possibly related/dup: #9413

**Comment by ComradeP** [comment]
For the record, this is still reproducible with version `1.6.0-rc7` from Debian's experimental archive.

**Comment by ComradeP** [comment]
Oddly, version `1.6.1` from Debian no longer shows this problem. Will test more extensively and close if confirmed.

**Comment by ComradeP** [comment]
Indeed, this seems to have been resolved in the meantime. Closing.

**Comment by ComradeP** [comment]
The exact same issue has resurfaced with the latest Debian version of 1.8.3, as well as with the latest version from the Docker-maintained repos (1.10.1). The underlying root filesystem is visible in running containers where it's not masked by stuff in the overlay filesystems, which both exposes information and breaks things (due to `/etc/ld.so.d` and others).

This is on the same machine as before, so the hardware setup is very similar. Here's the current diagnostic output

```
$ docker info
Containers: 31
 Running: 1
 Paused: 0
 Stopped: 30
Images: 1009
Server Version: 1.10.1
Storage Driver: btrfs
 Build Version: Btrfs v4.4
 Library Version: 101
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.3.0-1-amd64
Operating System: Debian GNU/Linux stretch/sid
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.07 GiB
Name: byakhee
ID: Q24M:KMC4:KYV3:SIQ4:TFAE:ABSM:7K5O:KENN:RGRN:DFZ2:4PS4:N25U
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No oom kill disable support
$ docker version
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:15:20 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:15:20 2016
 OS/Arch:      linux/amd64
$ uname -a
Linux byakhee 4.3.0-1-amd64 #1 SMP Debian 4.3.3-7 (2016-01-19) x86_64 GNU/Linux
```

Is there any other information that would be useful?

**Comment by thaJeztah** [comment]
I'll add this to the 1.10.2 milestone for investigation, let's see if we can reproduce

**Comment by thaJeztah** [comment]
@vbatts are you still using btrfs? Perhaps you have an idea?

**Comment by vbatts** [comment]
Oh wild. I've never experienced this before.

Let me get a few things clear in the set up.
Is the superblock of the host '/' the same as '/var/lib/docker'?
Is '/' of the host in its own subvolume? What about /var/lib/docker?
What kernel version and btrfs-progs version was used when the superblock was formatted(mkfs.btrfs)?

**Comment by ComradeP** [comment]
> Is the superblock of the host '/' the same as '/var/lib/docker'?

Sorry, I'm not quite sure what you mean. How do I find this out? `btrfs-show-superblock` works on devices rather than directories.

> Is '/' of the host in its own subvolume? What about /var/lib/docker?

Yes, [see comment above](https://github.com/docker/docker/issues/10216#issuecomment-93478594). `/` is a subvolume, but `/var/lib/docker` is just a subfolder in that subvolume, not its own subvolume.

> What kernel version and btrfs-progs version was used when the superblock was formatted(mkfs.btrfs)?

```
$ uname -a
Linux byakhee 4.3.0-1-amd64 #1 SMP Debian 4.3.3-7 (2016-01-19) x86_64 GNU/Linux
$ dpkg -l btrfs-tools                                                                                                                              ⏎
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                             Version               Architecture          Description
+++-================================-=====================-=====================-======================================================================
ii  btrfs-tools                      4.4-1                 amd64                 Checksumming Copy on Write Filesystem utilities
$ btrfs --version
btrfs-progs v4.4
```

**Comment by ComradeP** [comment]
> What kernel version and btrfs-progs version was used when the superblock was formatted(mkfs.btrfs)?

Sorry, I just realised you were asking about the version when the FS was created. According to the installer logs for the system, it used version `0.19+20120328-7.1` of `btrfs-tools-udeb` and version `3.2.57-3` of `btrfs-modules-3.2.0-4-amd64-di`, i.e. kernel version 3.2.0.

**Comment by j16sdiz** [comment]
I think btrfs have long standing issues with chroot / overlayfs. This combination should not be used.

See:
http://www.spinics.net/lists/linux-btrfs/msg51557.html
https://bugzilla.kernel.org/show_bug.cgi?id=101951

**Comment by thaJeztah** [comment]
ping @vbatts could you have another look? is this due to an old version used during creation?

**Comment by vbatts** [comment]
sorry. I had not had the bandwidth to setup a system that has the original versions used by @ComradeP mkfs.btrfs a partition setup to match, and attempt to reproduce that way.
I have had a btrfs partition that had been around for 10months through several updates, and seen nothing like this. Then I set up a vm that had this partition layout (but recent kernel and btrfs-progs version) and could not reproduce this issue either.

**Comment by vbatts** [comment]
@ComradeP This host of yours looks like a long established host, but my concern is that the early version of the tooling may have had a lasting impact on the integrity of the block device's metadata and data storage. I'm feeling it best to recommend a fresh reformat, with the newer versions of tooling.

Were `/var/lib/docker` in its own subvolume, there would be a way to migrate that subvolume to a new device, but as `/var/lib/docker` is just sitting on the host's `/` subvolume, I don't have a good recommendation to migrate to a new device.

There are other ways, but they are "creative" and delicate, so I will be slow to recommend them.

**Comment by ComradeP** [comment]
@vbatts -- hm, interesting. Following your comment of "Were `/var/lib/docker` in its own subvolume...", I tried creating a new subvolume in the same BTRFS filesystem and mounting it at `/var/lib/docker`. Lo and behold, the problems have gone away, and I can no longer see my own filesystem shining through into running containers.

This is probably enough of a workaround for me, but for the sake of helping the investigation, I'd like to point out that -- perhaps slightly unusually -- in my setup the root of the filesystem is not the root subvolume on BTRFS. Instead, the BTRFS partition contains subvolumes at `/@root`, `/@home` and (now) `/@docker`, which are mounted at `/`, `/home` and `/var/lib/docker` respectively. The BTRFS path `/` is not mounted anywhere by default.

**Comment by vbatts** [comment]
@ComradeP glad this is enough of a workaround for you. 

That does not sound like an unusual btrfs partition/subvolume layout to me. While I am just myself, and not a majority to be weighed against. Though I myself do a similar layout as you have described.

I will close this issue again, as it has not been seen by others, not able to be reproduced by others, and now has a workaround.

**Comment by skerit** [comment]
@vbatts Pardon my necrobump but I just had this issue too. With a single, specific image. No matter how many times I tried to create a new container with that image the host's root drive would get mounted.

I'm also using the BTRFS storage driver.
It's a pretty new system (only installed it last week)
The only "weird" thing I did was restore a snapshot of the root drive and then afterwards restore another snapshot.
